### PR TITLE
Add support to __half and nv_bfloat16 to most math functions

### PR DESCRIPTION
**Pull request by Kh4ster**
This PR adds support to __half and nb_bfloat16 to most math functions:

- cos
- sin
- exp
- log
- max
- min
- sqrt

### CODE REVIEW

File: cpp/include/raft/core/detail/macros.hpp [hunk 1]
```diff
@@ -22,6 +22,12 @@
 #endif
 #endif
 
+#if defined(_RAFT_HAS_CUDA)
+#define CUDA_ONLY_CONDITION(condition) condition
+#else
+#define CUDA_ONLY_CONDITION(condition) true
+#endif
```
Comment (Nyrio) [inline cpp/include/raft/core/detail/macros.hpp:29]: Do you mind future-proofing that with:
```c++
#if defined(_RAFT_HAS_CUDA)
#define CUDA_CONDITION_ELSE_TRUE(condition) condition
#define CUDA_CONDITION_ELSE_FALSE(condition) condition
#else
#define CUDA_CONDITION_ELSE_TRUE(condition) true
#define CUDA_CONDITION_ELSE_FALSE(condition) false
#endif
```

and using `CUDA_CONDITION_ELSE_TRUE`?

File: cpp/include/raft/core/math.hpp [hunk 2]
```diff
@@ -91,7 +96,14 @@ RAFT_INLINE_FUNCTION auto atanh(T x)
 
 /** Cosine */
 template <typename T>
-RAFT_INLINE_FUNCTION auto cos(T x)
+RAFT_INLINE_FUNCTION
+#if _RAFT_HAS_CUDA
+  typename std::enable_if<!std::is_same<T, __half>::value && !std::is_same<T, nv_bfloat16>::value,
+                          T>::type
```
Comment (Nyrio) [inline cpp/include/raft/core/math.hpp:102]: Nitpick: can you use C++17 constructs for readability? `enable_if_t` and `is_same_v` (everywhere, not just this line)

File: cpp/include/raft/core/math.hpp [hunk 3]
```diff
@@ -91,7 +96,14 @@ RAFT_INLINE_FUNCTION auto atanh(T x)
...
+#else
+  auto
+#endif
```
Comment (Nyrio) [inline cpp/include/raft/core/math.hpp:105]: Can you use `CUDA_ONLY_CONDITION` / `CUDA_CONDITION_ELSE_TRUE` when you have this kind of host guard?

File: cpp/include/raft/core/math.hpp [hunk 4]
```diff
@@ -100,9 +112,49 @@ RAFT_INLINE_FUNCTION auto cos(T x)
...
+{
+#if (__CUDA_ARCH__ >= 530)
+  return ::hcos(x);
+#else
+  // static_assert(false) would be evaluated during host compilation stage while __CUDA_ARCH__ is at
+  // device compilation stage Using this sizeof(T) != sizeof(T) makes it work as it's only triggered
+  // during template instantiation and thus at device compilation stage
+  static_assert(sizeof(T) != sizeof(T), "__half is only supported on __CUDA_ARCH__ >= 530");
+  return T{};
+#endif
```
Comment (Nyrio) [inline cpp/include/raft/core/math.hpp:128]: Question for @cjnolet : iirc RAPIDS is Pascal+, shall we have a centralized check of the minimum arch somewhere and remove these checks here, or would this have too many consequences on repos using raft?

File: cpp/include/raft/core/math.hpp [hunk 5]
```diff
@@ -151,16 +243,89 @@ RAFT_INLINE_FUNCTION auto exp(T x)
...
+template <typename T>
+RAFT_DEVICE_INLINE_FUNCTION typename std::enable_if<std::is_same<T, __half>::value, __half>::type
+exp(T x)
+{
+#if (__CUDA_ARCH__ >= 530)
+  return ::hexp(x);
+#else
+  // static_assert(false) would be evaluated during host compilation stage while __CUDA_ARCH__ is at
+  // device compilation stage Using this sizeof(T) != sizeof(T) makes it work as it's only triggered
+  // during template instantiation and thus at device compilation stage
```
Comment (Nyrio) [inline cpp/include/raft/core/math.hpp:256]: This comment appears in many places, can you please change it for a one-liner? Such as:
```c++
// Fail during template instantiation if the compute capability doesn't support this operation
```

or remove it altogether

**Review by Nyrio — COMMENTED** [review]
Thanks for the PR. A few nitpicks

File: cpp/include/raft/core/math.hpp [hunk 6]
```diff
@@ -100,8 +108,40 @@ RAFT_INLINE_FUNCTION auto cos(T x)
 #endif
 }
 
-/** Sine */
+#if _RAFT_HAS_CUDA
 template <typename T>
+RAFT_DEVICE_INLINE_FUNCTION typename std::enable_if<std::is_same_v<T, __half>, __half>::type cos(
```
Comment (Nyrio) [inline cpp/include/raft/core/math.hpp:113]: -> `enable_if_t`

_note: 18 occurences_

**Review by Nyrio — CHANGES_REQUESTED** [review]

File: python/pylibraft/pyproject.toml [hunk 7]
```diff
@@ -19,7 +19,7 @@ requires = [
     "cuda-python>=11.7.1,<12.0",
     "cython>=0.29,<0.30",
     "ninja",
-    "rmm==23.8.*",
+    "rmm==23.6.*",
```
Comment (Nyrio) [inline python/pylibraft/pyproject.toml:22]: Why is this required?

**Comment by Nyrio** [comment]
Note: this PR solves #1542

File: python/pylibraft/pyproject.toml (hunk 7 above)
Comment (Nyrio) [reply python/pylibraft/pyproject.toml]: @bdice Yes, I think the `dependencies.yaml` still has 23.6, so the right fix is to update that to 23.8

Comment (bdice) [reply python/pylibraft/pyproject.toml]: Fix is filed here: https://github.com/rapidsai/raft/pull/1556

**Review by Nyrio — APPROVED** [review]

**Review by cjnolet — APPROVED** [review]
LGTM. Thanks for the PR!
